modin/pandas/groupby.py

```diff
@@ -295,17 +304,130 @@ def apply(self, func, *args, **kwargs):
     def dtypes(self):
         if self._axis == 1:
             raise ValueError("Cannot call dtypes on groupby with axis=1")
+
+        def dtypes_getter(grp):
+            try:
+                return grp.dtypes
+            except AttributeError:
+                return grp.dtype
+
         if not self._as_index:
-            return self.apply(lambda df: df.dtypes)
+            return self.apply(dtypes_getter)
         else:
-            return self._apply_agg_function(lambda df: df.dtypes)
+            return self._apply_agg_function(dtypes_getter)
 
     def first(self, **kwargs):
         return self._default_to_pandas(lambda df: df.first(**kwargs))
 
     def backfill(self, limit=None):
         return self.bfill(limit)
 
+    _internal_by_cache = no_default
+
+    # TODO: since python 3.9:
+    # @cached_property
+    @property
+    def _internal_by(self):
+        """
+        Get only those components of 'by' that are column labels of the source frame.
+
+        Returns
+        -------
+        list of labels
+        """
+        if self._internal_by_cache is not no_default:
+            return self._internal_by_cache
```

Comment:
if we're certain that the structure of GroupBy object is constant it's better to cache this and the following properties, I haven't found any cases where the GroupBy objects get modified during their lifetime, so I assumed that we can cache it.

Are we solid about the constancy of GroupBy object?